lua/nvim-treesitter/query.lua

```diff
@@ -71,6 +72,70 @@ function M.get_query(lang, query_name)
   end
 end
 
+ --This function is copied from Neovims treesitter.lua to enforece same behavior
+local magic_prefixes = {['\\v']=true, ['\\m']=true, ['\\M']=true, ['\\V']=true}
+local function check_magic(str)
+  if string.len(str) < 2 or magic_prefixes[string.sub(str,1,2)] then
+    return str
+  end
+  return '\\v'..str
+end
+
+local function compile_regex(regex)
+  if not M._regex_cache then
+    M._regex_cache = {}
+  end
+  if not M._regex_cache[regex] then
+    local re = vim.regex(check_magic(regex))
+    M._regex_cache[regex] = re
+    return re
+  else
+    return M._regex_cache[regex]
+  end
+end
```

Comment:
Are we sure that regexes are not compiled in the query already ? If that's the case, then we should use it, if not, we should post an PR upstream to fix that.

Reply:
As far as I could see it it is performed only for `match?` right now (https://github.com/neovim/neovim/blob/bd5f0e9695cb21c8b97f844ce21432ee8d06b7ed/runtime/lua/vim/treesitter.lua#L140). Neovim/Treesitter do not know anything of the semantics of the directives so they cannot decide whether a string is a regex or something else.

Reply:
Oh you are right...
Then that is a problem, hen we will need for quite some time I guess...
On thing that might be worth investigating too (not really related, just want to throw that idea here) is trying to override `Query:match_preds` which is the part of the `Query` metatable which matches the predicates.

Another thing is maybe upstreaming some of our modifications to neovim, but that's another discussion.

Reply:
I think there's a difference between directives for choosing matches and application-specific annotations. I guess its even ok if a nvim-treesitter module has its own post-processing directives (they could then be implemented even in the module itself). Then, users can configure the behavior of the module on a query basis.

For textobjects it makes sense to trim/extend a matched node. Whitespaces will never be captured by the syntax tree but they are important for this feature. Also, to strip leading `// ` of comments when you want to change a inner comment.

Matching sub-regions of a node could also get relevant for highlighting of injected inline code in comments: You have to merge multiple comment nodes, remove leading `// ` and determine a region where the example code starts. We have to see which parts are useful in a greater context and which not.